File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>(); //create a list to store nodes in every level from left to right

      Queue<TreeNode> q = new LinkedList<>();
      //add root to the queue first
      if(root != null){
        q.add(root); 
      }

      while(!q.isEmpty()){
        List<Integer> level = new ArrayList<>();
        for(int i = q.size(); i > 0; i--){
                TreeNode curr = q.poll();
                if(curr != null){
                    level.add(curr.val);
                    q.add(curr.left);
                    q.add(curr.right);
                }
        }
        if(level.size() > 0){
            res.add(level);
        }
      }
        return res;

    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
diameter = num of edges of the longest path between any two nodes within the tree
for each node diameter
height left subtree + height of right subtree

"""
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0
        def height(node):
            nonlocal diameter
            if node is None:
                return 0 
            left_height = height(node.left)
            right_height = height(node.right)
            #At each node,  calculate the longest path passing through the current node
            diameter = max(diameter, left_height+right_height)
            #return height of substree
            return 1 + max(left_height,right_height)
        height(root)
        return diameter